src/rsz/src/SizeDownMove.cc

```diff
@@ -172,168 +171,271 @@ bool SizeDownMove::doMove(const Path* drvr_path,
+               network_->pathName(load_pin),
+               network_->pathName(output_pin),
+               load_cell->name(),
+               delayAsString(load_delay, sta_, 3),
+               output_load_cap,
+               delayAsString(output_slew, sta_, 3),
+               delayAsString(slack_margin, sta_, 3));
+  }
+
+  best_cell = load_cell;
+  float best_cap = load_input_cap;
+  float best_area = load_cell->area();
+  for (LibertyCell* swappable : swappable_cells) {
+    if (swappable == load_cell) {
+      continue;
+    }
+    debugPrint(logger_,
+               RSZ,
+               "size_down",
+               3,
+               " considering swap {} {} -> {}",
+               network_->pathName(load_pin),
+               load_cell->name(),
+               swappable->name());
+    LibertyPort* new_load_port = swappable->findLibertyPort(load_port_name);
+    float new_input_cap = new_load_port->cornerPort(lib_ap)->capacitance();
+    float new_area = swappable->area();
+
+    // Input cap and area improvement checking
+    // Note: we check area because we don't want a bigger cell with smaller
+    // cap.
+    if (new_input_cap > best_cap || new_area > best_area) {
+      debugPrint(logger_,
+                 RSZ,
+                 "size_down",
+                 3,
+                 "  skip based on cap/area {} gate={} cap={}>{} area={}>{}",
+                 network_->pathName(load_pin),
+                 swappable->name(),
+                 new_input_cap,
+                 best_cap,
+                 new_area,
+                 best_area);
+      continue;
     }
-    const Pin* fanout_output_pin = fanouts[max_fanout_id];
-    LibertyPort* fanout_output_port = network_->libertyPort(fanout_output_pin);
-    const char* fanout_output_port_name = fanout_output_port->name();
-
-    const float drvr_delay
-        = resizer_->gateDelay(drvr_port, drvr_load_cap, dcalc_ap);
-    const float fanout_delay
-        = resizer_->gateDelay(fanout_output_port, fanout_load_cap, dcalc_ap);
-    debugPrint(
-        logger_,
-        RSZ,
-        "size_down",
-        3,
-        "base delay {} FO={} drvr_delay={} fanout_delay={} fanout_slack={}",
-        network_->pathName(fanout_pin),
-        fanout_cell->name(),
-        delayAsString(drvr_delay, sta_, 3),
-        delayAsString(fanout_delay, sta_, 3),
-        delayAsString(fanout_slack, sta_, 3));
-
-    best_cell = fanout_cell;
-    float best_cap = fanout_cap;
-    float best_area = fanout_cell->area();
-    for (LibertyCell* swappable : swappable_cells) {
-      if (swappable == fanout_cell) {
-        continue;
+
+    // Max capacitance and slew checking
+    bool skip_cell = false;
+    for (int i = 0; i < output_pins.size(); i++) {
+      LibertyPort* output_port
+          = swappable->findLibertyPort(output_port_names[i]);
+
+      // Check for max capacitance violation
+      if (checkMaxCapViolation(output_pins[i], output_port, output_caps[i])) {
```

Comment:
I think this check suffers from corner confusion. Maybe leave a comment so we can eventually fix it together with other cases throughout the resizer